src/coreclr/jit/assertionprop.cpp

```diff
@@ -2238,18 +2260,24 @@ AssertionInfo Compiler::optAssertionGenJtrue(GenTree* tree)
             if (con >= 0)
             {
                 AssertionDsc dsc;
-
-                // For arr.Length != 0, we know that 0 is a valid index
-                // For arr.Length == con, we know that con - 1 is the greatest valid index
+                bool         useNextEdge;
                 if (con == 0)
                 {
+                    // arr.Length != 0 -> Then: idx = 0        Else: no assertion
+                    // arr.Length == 0 -> Then: no assertion,  Else: idx = 0
+                    //
                     dsc.assertionKind = OAK_NOT_EQUAL;
                     dsc.op1.bnd.vnIdx = vnStore->VNForIntCon(0);
+                    useNextEdge       = relop->OperIs(GT_EQ);
                 }
                 else
                 {
+                    // arr.Length == 10 -> Then: idx = [0..9],  Else: no assertion
+                    // arr.Length != 10 -> Then: no assertion,  Else: idx = [0..9]
+                    //
                     dsc.assertionKind = OAK_EQUAL;
                     dsc.op1.bnd.vnIdx = vnStore->VNForIntCon(con - 1);
+                    useNextEdge       = relop->OperIs(GT_NE);
```

Comment:
Where do we use these kinds of assertions? Can we change this to create an `OAK_SUBRANGE` assertion instead? The assertion created here just looks wrong to me.

Reply:
answered in https://github.com/dotnet/runtime/pull/110171#discussion_r1858168477

so we basically create a fake array access with constant index here. I do not see how `OAK_SUBRANGE ` can help here

Reply:
These assertions are only used by range check, and it does not look at the index part of these assertions, so the `con - 1` is just unused (and indeed meaningless, making the comment confusing). It would probably make more sense to add an `optOp1Kind` for an arbitrary type of expression here (that comes with no extra information). Or, if the subrange assertion has no diffs, that would also make sense to me.

I'm confused about the bug fixed in #110142, though. This code was creating assertion saying that `op1VN == 0`, which seems perfectly correct. So I don't understand what the bug was with using these assertions.

Reply:
> I'm confused about the bug fixed in https://github.com/dotnet/runtime/pull/110142, though. This code was creating assertion saying that op1VN == 0, which seems perfectly correct. So I don't understand what the bug was with using these assertions.

It is not correct, E.g. 

```
if (arr.Length == 1)
```

creates 

````
op1 = arr.Length
kind = OAK_EQUAL
op2.u1.icon = 0
```
assertion in the "then" edge which led to a bug, because op2.u1.icon is meaningless, but my logic took it as a value of arr.length